#150 - Multi-Phase Auction; Bidding Phase

bh2smith opened the pull request:
closes #150 
and also 
closes #153 

**Test Plan**: Unit tests

bh2smith committed:
Enable Auction Solution Bidding. Note that old `applyAuction` still remains.

bh2smith committed:
no need for only owner here.

bh2smith committed:
Including first half of unit tests (rejected transactions)

bh2smith committed:
lint solidity

bh2smith committed:
Final Test Cases

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -28,6 +28,14 @@ contract SnappAuction is SnappBase {
+        bytes32 tentativeState;
+    }
+
+    mapping (uint => AuctionResult) public auctionResults;
```
Can you comment what the key here is? Ideally this would be stored on the auction directly (assuming we refactor it to not use the PendingFlux struct).

bh2smith replied:
Yes. I was just about to address this `AuctionFlux` issue and noticed it would be absolutely necessary for the phase II anyway. However, I would personally prefer to merge this first and do the refactoring with the already existing auction struct.

fleupold replied:
Given that the current implementation needs fundamental changes in the way the phase checking is done, I'd prefer if we can only merge it once read. 

Are you worried about merge conflicts? I don't think you are touching any existing code here, so that should be fine.

bh2smith replied:
Yes, I can agree now that I saw the additional constraint regarding previous solution time. 

> I'd prefer if we can only merge it once read.

Not sure what this is.

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -187,6 +195,51 @@ contract SnappAuction is SnappBase {
+        // );
+
+        // Ensure that auction batch is inactive, unprocessed and in correct phase for bidding
+        require(auctions[slot].appliedAccountStateIndex == 0, "Auction already applied");  // This is a bit redundant.
```
How is this redundant? If the auction is between 3-6 minutes from starting it cannot be applied yet? Let's remove this check then.

bh2smith replied:
If I remember correctly, this is redundant because the auction could not have already been applied if it is in the correct phase for bidding. I had left it to make sure I hadn't missed any cases. We will find that this check is absolutely necessary during the second phase!

bh2smith replied:
Turns out it was not redundant and this was caught by our unit tests!

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -187,6 +195,51 @@ contract SnappAuction is SnappBase {
+
+        // Ensure that auction batch is inactive, unprocessed and in correct phase for bidding
+        require(auctions[slot].appliedAccountStateIndex == 0, "Auction already applied");  // This is a bit redundant.
+        require(slot != MAX_UINT && slot <= auctionIndex, "Requested order slot does not exist");
```
nit: auction slot instead of order slot.

bh2smith replied:
```suggestion
        require(slot != MAX_UINT && slot <= auctionIndex, "Requested auction slot does not exist");
```

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -187,6 +195,51 @@ contract SnappAuction is SnappBase {
+        require(
+            block.timestamp > auctions[slot].creationTimestamp + 3 minutes ||
+                auctions[slot].size == maxUnreservedOrderCount(),
+            "Requested order slot is still active"
```
ditto: auction slot

bh2smith replied:
This is supposed to say order slot. Unless you simply prefer it to say auction slot... hmmm maybe I guess if by auction slot you mean the entire collection of orders.. then I guess. Will change.

fleupold replied:
I think auction slot is a better name. Especially given that orders might in the future span multiple auctions.

bh2smith replied:
Note that there are several other places in our code that have this same mistake.

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -187,6 +195,51 @@ contract SnappAuction is SnappBase {
+        require(slot != MAX_UINT && slot <= auctionIndex, "Requested order slot does not exist");
+        require(
+            block.timestamp > auctions[slot].creationTimestamp + 3 minutes ||
+                auctions[slot].size == maxUnreservedOrderCount(),
```
closing an auction early because of too many orders will result in a longer bidding period? As in if we close the batch 2 minutes after creation, we will allow a bidding period for 4 minutes?

I think this is fine for now as we will hopefully get rid of the limited order space soon.

bh2smith replied:
Yes. this certainly an item of interest! Good find. I had also sketched this up yesterday and meant to include it in the "Problems" section of our daily standup. I don't think its a terrible thing to allow more solving time when the batch is completely full!

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -187,6 +195,51 @@ contract SnappAuction is SnappBase {
+        bytes32 _currStateRoot,
+        bytes32 proposedStateRoot,
+        uint proposedObjectiveValue
+        // bytes32 _orderHash,
```
Yes, let's please remove all the commented code.

fleupold reviewed (commented):
Question regarding the phases: Assume we have two auctions

```
A1:  | order_collection | solution bidding | solution posting
A2:                     | order collection | solution bidding |  solution posting
```

We require that *solution bidding* of A2 can only happen after the *solution posting* of A1 has been completed (which makes sense since people wouldn't be able to know the balances of all accounts in A2 before seeing what trades happened in A1).

However, this also means that the solver of A1 can wait until the end of their *solution posting* period and thus cut short the *solution bidding* period of A2 for everyone else (they themselves could have started solving A2 as soon as *order collection* was over).

This seems unfair to other competitors that try to solve A2.

I'd suggest that the *solution bidding* only starts once the result of the previous auction is revealed (and even more order collection only ends at this this point).

Therefore I think we should store the time of the solution submission next to the appliedStateRoot (requires refactoring PendingFlux into a separate PendingAuction) and do the phase evaluation based on the `max(auctionCreationTime + 3 minutes, previousAuctionSolutionRevelationTime)`

bh2smith committed:
Update contracts/SnappAuction.sol

bh2smith committed:
Suggestions:
- rename order slot to auction slot in appropriate places
- introduce new struct AuctionBatch
- delete commented code for order hash verification

bh2smith committed:
typo in Pending Batch struct comments.

bh2smith committed:
include solutionAcceptedTime in Pending Auction.

bh2smith committed:
include new field in empty Batch initialization.

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -26,7 +26,20 @@ contract SnappAuction is SnappBase {
+        bytes32 solutionHash;          // Succinct record of trade execution & prices
+        uint solutionAcceptedTime;   // Time solution was accepted (written at time of solutionHash)
+        bytes32 tentativeState;        // Proposed account state during bidding phase
+        uint16 size;                   // Number of orders in this auction
```
ubernit: if small variables are at all together (at then end of the struct), solidity can collate them into a single word.

Maybe we can also call this something like number of orders?

bh2smith replied:
I don't understand your ubernit.

I wont change the `size` here, because I will like use the `PendingBatch` struct from the `SnappBase`

josojo replied:
Why do we need to store the size of orders?
I think in all fraud proofs, we can deduct the size from the solution submission.

bh2smith replied:
The size is needed in order to emit the order index (i.e. so the contract knows which order it is).

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -26,7 +26,20 @@ contract SnappAuction is SnappBase {
+        uint solutionAcceptedTime;   // Time solution was accepted (written at time of solutionHash)
+        bytes32 tentativeState;        // Proposed account state during bidding phase
+        uint16 size;                   // Number of orders in this auction
+        bytes32 shaHash;               // Rolling shaHash of all orders
```
nit: let's call this order hash?

fleupold replied:
Actually looking at this now, it might be equally good to reuse the PendingBatch struct nested as you suggested over slack. Up to you. Having nested structs could serve as a more structured documentation. Something like

```
struct PendingAuction {
  PendingBatch orderBatch,
  SolutionBid bestSolution
  uint solutionRevealedTime
}
```

bh2smith replied:
I'll give it a look! Turns out there is a loss of coverage because of a new if branch that I need to deal with anyway.

bh2smith replied:
Unfortunately we can't have nested structs at this point (as it is still Experimental). I got this

```
The following types are only supported for getters in the new experimental ABI encoder:
struct SnappBaseCore.PendingBatch memory. 
Either remove "public" or use "pragma experimental ABIEncoderV2;" to enable the feature.
```

bh2smith replied:
So then I will be calling it `orderHash` as suggested.

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -187,6 +200,48 @@ contract SnappAuction is SnappBase {
+        require(auctions[slot].appliedAccountStateIndex == 0, "Auction already applied");
+        require(slot != MAX_UINT && slot <= auctionIndex, "Requested auction slot does not exist");
+
+        uint biddingStartTime = auctions[slot].creationTimestamp + 3 minutes;
```
Can we add a comment here explaining why we do this?

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -187,6 +200,48 @@ contract SnappAuction is SnappBase {
+            "Requested auction slot is still active"
+        );
+        require(
+            block.timestamp < auctions[slot].creationTimestamp + 6 minutes,
```
This should be biddingStartTime + 3 minutes, shouldn't it?

bh2smith replied:
Yes sir it should!

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -26,7 +26,20 @@ contract SnappAuction is SnappBase {
+        address solver;
+        uint objectiveValue;           // Traders utility
+        bytes32 solutionHash;          // Succinct record of trade execution & prices
+        uint solutionAcceptedTime;   // Time solution was accepted (written at time of solutionHash)
```
How about fullSolutionRevealedTime?

nit: Are the variable collated in some meaningful way? We could order it in the order that they are used (1st all order related stuff, then solution bidding, then solution applied)?

bh2smith replied:
Unfortunately, this is not the `fullSolutionRevealed` time because of the second phase. There is a difference between when the solution is revealed and when the solution is accepted. For more information on this, please refer to the discussion #151 (specifically the most recent post about `fallbackApplyAuction`).

Yes, I will try and order the variables better.

fleupold reviewed (approved):
Looks really good. Some smaller comments around naming inline.

josojo reviewed (approved):
Looks really good. 

You might want to add the bonding, as discussed in the issue.

bh2smith committed:
reorder and rename PendingAuction fields

bh2smith committed:
rearrange PendingAuction field initilization and some unit test typos

bh2smith committed:
Full Coverage - special case for biddingStartTime

bh2smith commented:
> You might want to add the bonding, as discussed in the issue.

Yes @josojo, I do want to add this, but think this will be better to include once the second phase has been implemented (i.e. when the bond can be returned).

bh2smith committed:
insert small comment.

bh2smith committed:
MOve the last waited second above applyAuction!

bh2smith committed:
final fix: wait 1 second between applyAuction and auctionSolutionBid

bh2smith commented:
Although this is already merged, I might prefer to rename `PendingAuction` to `AuctionBatch` (since the struct contains information about all three-phases of the auction). Do either of @josojo or @fleupold have any comments.

fleupold commented:
I like `PendingAuction` more than `AuctionBatch`. Shouldn't it be at least `BatchAuction` or maybe `PendingBatchAuction`?

bh2smith commented:
Yes, something like that, but it contains more information than simply the pending information. It literally is the entire auction (during all of its phases). For now, I will leave it... but of the things you suggested, I like `BatchAuction` best!
